ampd/src/tx_broadcaster/msg_queue.rs

```diff
@@ -0,0 +1,512 @@
+    }
+}
+
+impl MsgQueue {
+    pub fn new_msg_queue_and_client<T, A>(
+        cosmos_client: T,
+        account_manager: A,
+        msg_cap: usize,
+        gas_cap: Gas,
+        duration: time::Duration,
+    ) -> Result<(impl Stream<Item = Vec<Msg>>, MsgQueueClient<T, A>)>
+    where
+        T: cosmos::CosmosClient,
+        A: account::AccountManager,
+    {
+        let (tx, rx) = mpsc::channel(msg_cap);
+
+        Ok((
+            Box::pin(MsgQueue {
+                stream: ReceiverStream::new(rx).fuse(),
+                deadline: None,
+                queue: Queue::default(),
+                gas_capacity: gas_cap,
+                duration,
+            }),
+            MsgQueueClient {
+                tx,
+                cosmos_client,
+                account_manager,
+            },
+        ))
+    }
+}
+
+impl Stream for MsgQueue {
+    type Item = Vec<Msg>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut me = self.as_mut().project();
+
+        loop {
+            match me.stream.as_mut().poll_next(cx) {
+                Poll::Pending => break,
+                Poll::Ready(Some(msg)) => {
+                    if me.queue.is_empty() {
+                        me.deadline.set(Some(time::sleep(*me.duration)));
+                    }
+
+                    // gas overflow
+                    if me.queue.gas_cost().checked_add(msg.gas).is_none() {
+                        let msgs = me.queue.pop_all();
+                        me.queue.push(msg).expect("gas must not overflow");
+
+                        return Poll::Ready(Some(msgs));
+                    } else {
+                        me.queue.push(msg).expect("gas must not overflow");
+                    };
+
+                    if me.queue.gas_cost() >= *me.gas_capacity {
+                        return Poll::Ready(Some(me.queue.pop_all()));
+                    }
+                }
+                Poll::Ready(None) => {
+                    // Returning Some here is only correct because we fuse the inner stream.
+                    let last = if me.queue.is_empty() {
+                        None
+                    } else {
+                        Some(me.queue.pop_all())
+                    };
+
+                    return Poll::Ready(last);
+                }
+            }
+        }
+
+        if me.queue.is_empty() {
+            return Poll::Pending;
+        }
+
+        if let Some(deadline) = me.deadline.as_pin_mut() {
+            ready!(deadline.poll(cx));
+        }
+
+        Poll::Ready(Some(me.queue.pop_all()))
+    }
+}
+
+#[derive(Default)]
+struct Queue {
+    msgs: Vec<Msg>,
+    gas_cost: Gas,
+}
+
+impl Queue {
+    pub fn push(&mut self, msg: Msg) -> Result<()> {
+        self.gas_cost = self
+            .gas_cost
+            .checked_add(msg.gas)
+            .ok_or(report!(Error::IntegerOverflow))?;
+        self.msgs.push(msg);
+
+        Ok(())
+    }
```

Comment:
could we incorporate the `pop_all` here? I.e. if the gas would overflow, pop and return existing messages, else just psuh the current one. Then the calling code doesn't have to worry about the queues internal consistency and can just handle the two cases where it either gets nothing or the previous messages back